libsolidity/formal/CHC.cpp

```diff
@@ -839,9 +839,10 @@ void CHC::internalFunctionCall(FunctionCall const& _funCall)
 	connectBlocks(
 		m_currentBlock,
 		predicate(*m_errorDest),
-		errorFlag().currentValue() > 0
+		errorFlag().currentValue() > 0 && currentPathConditions()
 	);
-	m_context.addAssertion(errorFlag().currentValue() == 0);
+	m_context.addAssertion(smtutil::Expression::implies(currentPathConditions(), errorFlag().currentValue() == 0));
```

Comment:
Why do we need the implication here?

Reply:
This one is interesting.
We need the implication, otherwise the branch that is not taken would influence what is possible in the branch taken.
It would forbid some states in the branch taken which would be allowed.
The example I had for that is something like this:
```
contract C {

	function decrement(uint x) private pure returns (uint) {
		return x - 1;
	}

	function decrement2(uint x) private pure returns (uint) {
		return x - 1;
	}

	function f(uint x) public pure returns (uint) {
		return x > 0 ? decrement(x) : decrement2(x);
	}
}
```
Without the conditional in the code, SMTChecker would not discover the underflow in the second branch.
But I will double check this.

Reply:
I checked it and it is as I said above. Should I add that example to the tests?

Reply:
Ok, makes sense. But why isn't the same implication needed for external calls?
Yes please add the test.

Reply:
I think it might be needed. But to deal with the external calls properly, we need to first resolve https://github.com/ethereum/solidity/issues/14111.
I would propose to first close this one, then look at the external call issue and only after that revisit the external calls inside ternary operator.

Reply:
The test has been added and PR has been rebased.